core/src/program.rs

```diff
@@ -256,22 +312,24 @@ impl<EC: EvalCache> Program<EC> {
     ) -> std::io::Result<Self>
     where
         T: Read,
-        S: Into<OsString> + Clone,
+        S: Into<OsString>,
     {
-        increment!("Program::new");
-
-        let mut cache = Cache::new(ErrorTolerance::Strict);
-        let path = PathBuf::from(source_name.into());
-        let main_id = cache.add_source(SourcePath::Path(path), source)?;
-        let vm = VirtualMachine::new(cache, trace);
+        Self::new_from_input(Input::Source(source, source_name.into()), trace)
```

Comment:
It seems we are always coercing `source_name` to an `OsString` before putting it in the `Input::Source`. So is it worth it to have the additional type parameter `S` on `Input`? Couldn't we just replace it with `OsString` without having to change anything else?